File: python-client/ds_rpc.py
import json

import requests
from python import deserialize_results, serialize_python_function

SETTINGS_FILE = 'settings.json'
VERSION_NUMBER = 0.1


class Server(object):

    def __init__(self):

        settings = json.load(open(SETTINGS_FILE, 'r'))

        # required settings
        try:
            self.url = settings['server_url']
        except KeyError:
            raise KeyError('Could not load server url from settings file')

        self.serializer = serialize_python_function

    def python(self, _function, *args):
        """Call a python function remotely
        Args:
            _function (str): function to call
            args (tuple): arguments to be passed to the given function

        Returns function result
        """

        return self._remote_call('python', _function, args)

    def r(self, _function, *args):
        """Call an R function remotely"""

        return self._remote_call('r', _function, args)

    def _remote_call(self, language, _function, args):
        """Execute remote call"""
        data = {
            'version': VERSION_NUMBER,
            'language': language
        }
        function_data = self.serializer(_function, args)
        data.update(function_data)
        response = requests.post(self.url, data=json.dumps(data))
        return deserialize_results(response.text)

File: server/js-client/ds_rpc.js

var VERSION_NUMBER = 0.1;
function Server(url){
    this.url = url;
}
var res;

Server.prototype.remote_call = function(callback, language, _function, args){
    serialized_call = {
        'version': VERSION_NUMBER,
        'language': language,
        '_function': _function,
        'n_args': args.length - 2
    };

    for (var i = 1; i < args.length; i ++){
        serialized_call['arg' + (i-2)] = args[i];
    }

    var request = $.ajax({
      url: "/",
      type: "POST",
      dataType:'json',
      ContentType: 'application/json',
      data: JSON.stringify(serialized_call),
      success: function(response){
        var data = [];
        for(var key in response) {
            data.push(response[key]);
        }
        callback(data);
      }
    });
};

Server.prototype.python = function(callback, _function){
    return this.remote_call(callback, 'python', _function, arguments);
};

Server.prototype.r = function(callback, _function){
    return this.remote_call(callback, 'r', _function, arguments);
};


